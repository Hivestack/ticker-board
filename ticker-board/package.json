{
  "name": "ticker-board",
  "version": "1.0.1",
  "description": "A JavaScript plugin for creating a tickerboard effect.",
  "main": "src/index.js",
  "scripts": {
    "build": "webpack --mode=production",
    "dev": "webpack --mode=development --watch",
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/rjkerrison/ticker-board.git"
  },
  "keywords": [
    "tickerboard",
    "train",
    "departures",
    "animation",
    "text",
    "effect"
  ],
  "author": "Robin James Kerrison",
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/rjkerrison/ticker-board/issues"
  },
  "homepage": "https://github.com/rjkerrison/ticker-board#readme",
  "devDependencies": {
    "css-loader": "^5.2.6",
    "style-loader": "^2.0.0",
    "url-loader": "^4.1.1",
    "webpack": "^5.40.0",
    "webpack-cli": "^4.7.2"
  }
}
