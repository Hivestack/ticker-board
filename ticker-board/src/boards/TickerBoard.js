class TickerBoard {
  constructor(selector, { size }) {
    const selectedElements = document.querySelectorAll(selector)
    this.boards = Array.from(selectedElements).map((selected) => {
      const messages = Array.from(selected.children).map((m) => m.innerText)
      const boardElement = document.createElement('ul')
      selected.replaceWith(boardElement)

      const board = new RotationBoard(boardElement, {
        count: messages.length,
        size: size || Math.max(...messages.map((x) => x.length)),
        messages,
      })

      boardElement.addEventListener('click', () => board.advance())

      return board
    })
  }

  advance(index) {
    this.boards[index].advance()
  }
}

export default TickerBoard
