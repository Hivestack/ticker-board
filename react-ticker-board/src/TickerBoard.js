import React, { useEffect, useRef, useState, useCallback } from 'react'
import { Board } from 'ticker-board'

let idCounter = 1

const TickerBoard = ({ messages, count, size, delay, theme }) => {
  const [board, setBoard] = useState(null)
  const [element, setElement] = useState(null)

  const id = useRef(idCounter++)

  const ref = useCallback((node) => {
    if (node !== null && node !== element) {
      setElement(node)
    }
  }, [])

  useEffect(() => {
    if (element && !board) {
      const board = new Board(element, { count, size, delay, theme })
      setBoard(board)
    }
  }, [element])

  useEffect(() => {
    if (board) {
      board.updateCount(count)
      // Important: we need to keep messages immutable. We'll pass a copy, not the reference.
      board.updateMessages([...messages])
    }
  }, [board, messages, count])

  return (
    <div id="ticker">
      {/*
        The ref is a child so that it can be freely manipulated by Board.
      Â  React will unmount at the higher level.
      */}
      <div id="ticker-child" ref={ref}>
        <div></div>
        <span data-testid="instance-id">{id.current}</span>
      </div>
    </div>
  )
}

export default TickerBoard
