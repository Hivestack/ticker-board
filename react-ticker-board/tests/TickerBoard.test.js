import React from 'react'
import { render, screen } from '@testing-library/react'

jest.mock('ticker-board')

import TickerBoard from '../src/TickerBoard'
import { Board } from 'ticker-board'

const messages = ['Example', 'testing', 'messages']
const count = messages.length

let node
let updateMessages
let updateCount
let component

beforeEach(() => {
  node = document.createElement('div')
  updateMessages = jest.fn()
  updateCount = jest.fn()

  Board.mockReturnValue({
    ...Board,
    updateMessages,
    updateCount,
  })

  component = render(
    <TickerBoard messages={messages} count={count} size={24} theme={'dark'} />,
    node
  )
})

describe(TickerBoard, () => {
  it('renders', () => {
    expect(component.baseElement).toMatchSnapshot('TickerBoard renders')
  })

  it('displays the messages', () => {
    // ASSERT
    // Check updateMessages is called as we'd expect
    expect(updateMessages).toHaveBeenCalledTimes(1)
    expect(updateMessages).toHaveBeenCalledWith(messages)
  })

  it('updates the messages', () => {
    const { rerender } = component
    const timesCalled = Board.mock.calls.length
    const id = screen.getByTestId('instance-id').textContent

    const newMessages = ['Second set', 'of messages']
    rerender(
      <TickerBoard
        messages={newMessages}
        count={count}
        size={24}
        theme={'dark'}
      />,
      node
    )

    // ASSERT
    // no new board on rerender
    expect(Board).toHaveBeenCalledTimes(timesCalled)
    // and no new TickerBoard component
    expect(screen.getByTestId('instance-id')).toHaveTextContent(id)

    // Check updateMessages is called an extra time as we'd expect
    expect(updateMessages).toHaveBeenCalledTimes(2)
    expect(updateMessages).toHaveBeenCalledWith(newMessages)
  })
})
