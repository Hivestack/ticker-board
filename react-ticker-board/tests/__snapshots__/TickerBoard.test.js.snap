// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TickerBoard renders: TickerBoard renders 1`] = `
<body>
  <div>
    <div
      id="ticker"
    >
      <div
        id="ticker-child"
      >
        <div />
        <span
          data-testid="instance-id"
        >
          1
        </span>
      </div>
    </div>
  </div>
</body>
`;
